src/transport.rs

```diff
@@ -428,16 +410,61 @@ impl WebRTCDirectTransport {
             .map_err(Error::WebRTC)
             .await?;
 
+        // Create a datachannel with label 'data'
+        let data_channel = peer_connection
+            .create_data_channel(
+                "data",
+                Some(RTCDataChannelInit {
+                    id: Some(1),
+                    ..RTCDataChannelInit::default()
+                }),
+            )
+            .await?;
+
+        let (tx, mut rx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+        // Wait until the data channel is opened and detach it.
+        crate::connection::register_data_channel_open_handler(data_channel, tx).await;
+
         // Wait until data channel is opened and ready to use
-        select! {
-            res = data_channel_tx => match res {
-                Ok(dc) => Ok(Connection::new(peer_connection, dc)),
-                Err(e) => Err(Error::InternalError(e.to_string())),
+        let detached = select! {
+            res = rx => match res {
+                Ok(detached) => detached,
+                Err(e) => return Err(Error::InternalError(e.to_string())),
             },
-            _ = Delay::new(Duration::from_secs(10)).fuse() => Err(Error::InternalError(
+            _ = Delay::new(Duration::from_secs(10)).fuse() => return Err(Error::InternalError(
                 "data channel opening took longer than 10 seconds (see logs)".into(),
             ))
+        };
+
+        trace!("noise handshake with {}", remote);
+        let dh_keys = Keypair::<X25519Spec>::new()
+            .into_authentic(&self.id_keys)
+            .unwrap();
+        let noise = NoiseConfig::xx(dh_keys);
+        let info = noise.protocol_info().next().unwrap();
+        // after noise is successful and we've authenticated the remote peer, encrypted IO is no
+        // longer needed, hence ignored here.
+        let (peer_id, _) = noise
+            .upgrade_outbound(PollDataChannel::new(detached), info)
+            .and_then(|(remote, io)| match remote {
+                RemoteIdentity::IdentityKey(pk) => future::ok((pk.to_peer_id(), io)),
+                _ => future::err(NoiseError::AuthenticationFailed),
+            })
+            .await
+            .map_err(Error::Noise)?;
+
+        // Verify peer's identity.
+        let peer_id_from_addr = PeerId::try_from_multiaddr(&addr);
+        if peer_id_from_addr.is_none() || peer_id_from_addr.unwrap() != peer_id {
+            error!(
+                "peer_id_from_addr {:?} != peer_id {:?}",
+                peer_id_from_addr, peer_id
+            );
+            return Err(Error::InvalidMultiaddr(addr.clone()));
         }
+
+        Ok((peer_id, Connection::new(peer_connection).await))
```

Comment:
Is this sufficient to authenticate the remote and be able to fall back to the underlying WebRTC connection afterwards? Would one not need to exchange the TLS certificate fingerprint in order to prevent a monster-in-the-middle attack on the DTLS secured connection?

Reply:
> Is this sufficient to authenticate the remote and be able to fall back to the underlying WebRTC connection afterwards? Would one not need to exchange the TLS certificate fingerprint in order to prevent a monster-in-the-middle attack on the DTLS secured connection?

are you asking if we need to verify TLS certificate fingerprints (which we've verified during the DTLS handshake earlier) on top of Noise? Not a security expert here, but even if someone is trying to MiMA DTLS connection (which is possible), I can't see how they will be able to fake `peer_id`. So if our identity system is based on `peer_id`, then I believe we can be certain that the peer is the owner of public key associated with given `peer_id`.

Reply:
Say that _A_ wants to connect to _B_. _B_ advertises the `Multiaddr` (`/ip6/<b-ip>/udp/xxx/webrtc/<b-hash>/xxx/p2p/<b-peer-id>`).

Eve intercepts B's `Multiaddr` and forwards an altered `Multiaddr` to _A_, namely `/ip6/<e-ip>/udp/xxx/webrtc/<e-hash>/xxx/p2p/<b-peer-id>` (note `e-ip` and `e-hash`).

_A_ establishes a WebRTC connection to _E_ thinking it is talking to _B_. On the established connection _A_ initiates a Noise handshake. _E_ forwards the noise handshake payloads to _B_. Once the Noise handshake succeeds _A_ falls back to the underlying WebRTC connection, thinking it is talking to _B_ instead of _E_.

By exchanging the TLS certificate fingerprints in the Noise handshake, the above attack is prevented.

Does this make sense @melekes?

Reply:
It does 👍 Good thinking. I will make the necessary changes